File: problems/20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        parentheses = {
            "}" : "{",
            ")" : "(",
            "]" : "["
        }

        stack = []
        for p in s:
            if p in parentheses.values():
                stack.append(p)
            else:
                # if the stack is not empty. 
                # In Python, an empty list is considered False, while a non-empty list is considered True.
                if stack and stack[-1] == parentheses[p]:
                    stack.pop()
                else:
                    return False
        
        return len(stack) == 0

File: problems/20-valid-parentheses/valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
const isValid = (s) => {
    const parentheses = {
        ')': '(',
        '}': '{',
        ']': '[' 
    };

    let stack = [];
    for (let char of s) {
        if (Object.values(parentheses).indexOf(char) !== -1) {
            stack.push(char);
        } else {
            if (stack[stack.length - 1] === parentheses[char]) {
                stack.pop()
            } else {
                return false;
            }
        }
    }
    return stack.length === 0;
};